### ISSUE TITLE
login try limit agains bruteforce attack

### ISSUE THREAD

**Issue by usern283** [gogits/gogs#2291]

**Comment by yurivict** [comment]
This is generally hard to do because the same brute-force attack can come from different IP addresses. If attackers use a bot, each hit will come from the different IP. If you will simply delay the login, such defense can cause anybody to lock anybody else out of the system by continuous brute-forcing.

Better use 2-factor authentication, like google two-factor authenticator through PAM.

**Comment by usern283** [comment]
Yes, but I think that a brute-force attack is more difficult , if only 5 login attempt per IP are possible.

**Comment by initbar** [comment]
I agree with @yurivict's insight: rather than placing an overhead of n-logins by the unique IP set (and flushing the log once the user successfully logs in), I think the delay with exponential growth time is more elegant solution.

**Comment by stonerl** [comment]
Does Gogs currently log failed login attempts with the corresponding IP-Address? If so fail2ban could be used to take care of this.

**Comment by yurivict** [comment]
Ban who, user-id? IP? Net?

**Comment by stonerl** [comment]
@yurivict fail2ban bans IPs

**Comment by stonerl** [comment]
In general I don't think that it is Gogs responsibility to to take care of this matter, there are tools like fai2ban that are designed to do exactly this single job. Gogs just needs to provide a sane log file that allows such tools to work properly.

**Comment by yurivict** [comment]
agree. should just integrate gogs with fail2ban.

**Comment by stonerl** [comment]
I just checked the logfiles. Currently gogs does not log failed login attempts.

**Comment by ghevge** [comment]
Any updates on this topic? Thanks
